File: src/renderer/vulkan/PipelineFactory.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2023 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "PipelineFactory.h"

#include "Backend.h"
#include "core/Assert.h"

namespace Reaper
{
PipelineFactory create_pipeline_factory(VulkanBackend& backend)
{
    static_cast<void>(backend);

    return {
        .trackers = {},
        .pipelines = {},
        .dirty = true,
    };
}

void destroy_pipeline_factory(VulkanBackend& backend, PipelineFactory& pipeline_factory)
{
    for (auto pipeline : pipeline_factory.pipelines)
    {
        vkDestroyPipeline(backend.device, pipeline, nullptr);
    }

    pipeline_factory.pipelines.clear();
}

u32 register_pipeline_creator(PipelineFactory& factory, const PipelineCreator& creator)
{
    const u32 index = static_cast<u32>(factory.trackers.size());

    factory.trackers.emplace_back(PipelineTracker{
        .creator = creator,
        .loaded_version = 0,
        .pipeline_index = PipelineTracker::InvalidIndex,
    });

    return index;
}

namespace
{
    VkPipeline create_new_pipeline_from_tracker(VkDevice device, PipelineTracker& tracker,
                                                ShaderModules& shader_modules)
    {
        // tracker
        VkPipeline pipeline =
            tracker.creator.pipeline_creation_function(device, shader_modules, tracker.creator.pipeline_layout);

        tracker.loaded_version += 1;

        return pipeline;
    }
} // namespace

void pipeline_factory_update(VulkanBackend& backend, PipelineFactory& pipeline_factory, ShaderModules& shader_modules)
{
    if (pipeline_factory.dirty)
    {
        for (auto& tracker : pipeline_factory.trackers)
        {
            if (tracker.loaded_version == 0)
            {
                VkPipeline pipeline = create_new_pipeline_from_tracker(backend.device, tracker, shader_modules);

                tracker.pipeline_index = static_cast<u32>(pipeline_factory.pipelines.size());

                pipeline_factory.pipelines.push_back(pipeline);
            }
        }

        pipeline_factory.dirty = false;
    }
}

VkPipeline get_pipeline(const PipelineFactory& pipeline_factory, u32 pipeline_tracker_index)
{
    const PipelineTracker& tracker = pipeline_factory.trackers[pipeline_tracker_index];
    Assert(tracker.loaded_version != 0);

    return pipeline_factory.pipelines[tracker.pipeline_index];
}
} // namespace Reaper

File: src/renderer/vulkan/pipeline_factory.zig
// Port of src/renderer/vulkan/PipelineFactory.{h,cpp}
//
// Passes register a creator up front and get back a tracker index; the actual
// VkPipeline is built lazily on the first update. The indirection exists so
// pipelines can be rebuilt when shaders change — the version counter is the
// hook for that, even though nothing bumps it yet.
//
// The C++ takes a function pointer; here it is a function pointer too rather
// than a closure, so the registry stays a plain array with no captured state.

const std = @import("std");
const vk = @import("vulkan");

pub const PipelineFunctor = *const fn (
    vkd: *const vk.DeviceWrapper,
    device: vk.Device,
    pipeline_layout: vk.PipelineLayout,
) anyerror!vk.Pipeline;

pub const PipelineCreator = struct {
    pipeline_layout: vk.PipelineLayout,
    pipeline_creation_function: PipelineFunctor,
};

pub const PipelineTracker = struct {
    pub const invalid_index: u32 = 0xFFFF_FFFF;

    creator: PipelineCreator,
    loaded_version: u32 = 0,
    pipeline_index: u32 = invalid_index,
};

pub const PipelineFactory = struct {
    trackers: std.ArrayList(PipelineTracker) = .empty,
    pipelines: std.ArrayList(vk.Pipeline) = .empty,
    dirty: bool = true,

    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) PipelineFactory {
        return .{ .allocator = allocator };
    }

    pub fn deinit(self: *PipelineFactory, vkd: anytype, device: vk.Device) void {
        for (self.pipelines.items) |pipeline| {
            vkd.destroyPipeline(device, pipeline, null);
        }

        self.pipelines.deinit(self.allocator);
        self.trackers.deinit(self.allocator);
    }

    /// Returns a tracker index.
    pub fn registerPipelineCreator(self: *PipelineFactory, creator: PipelineCreator) !u32 {
        const index: u32 = @intCast(self.trackers.items.len);

        try self.trackers.append(self.allocator, .{ .creator = creator });

        return index;
    }

    /// Builds every pipeline that has not been built yet. Called once per frame
    /// before recording, same as the C++.
    pub fn update(self: *PipelineFactory, vkd: *const vk.DeviceWrapper, device: vk.Device) !void {
        if (!self.dirty) return;

        for (self.trackers.items) |*tracker| {
            if (tracker.loaded_version != 0) continue;

            const pipeline = try tracker.creator.pipeline_creation_function(
                vkd,
                device,
                tracker.creator.pipeline_layout,
            );

            tracker.loaded_version += 1;
            tracker.pipeline_index = @intCast(self.pipelines.items.len);

            try self.pipelines.append(self.allocator, pipeline);
        }

        self.dirty = false;
    }

    pub fn getPipeline(self: *const PipelineFactory, pipeline_tracker_index: u32) vk.Pipeline {
        const tracker = self.trackers.items[pipeline_tracker_index];
        std.debug.assert(tracker.loaded_version != 0);

        return self.pipelines.items[tracker.pipeline_index];
    }
};
